### PR TITLE
lodash exercises

### PR DESCRIPTION
**Pull request by amandanagai**
- not sure how to only push lodash and not also recursion exercises from yesterday (have not made any significant updates to the latter)
- for lodash, didn't get: includes (spent well over an hour on it today...probably should have moved on earlier, as you'd suggested), cloneDeep (think it has to be recursive, but am not sure what the base case is...), and padEnd
- also, I think overall, I'm doing and thinking too much in the imperative
- moved too quickly for me in class when we got small problems and you said: solve, imperative, declarative, recursively. I think that's the work I really need to do, but I'm having trouble ramping my brain up fast enough to do it once before we move on to the next 
- stopping b4 last 3 lodash functions are done bc I need to sleep or attempt to

### CODE REVIEW

File: canvas_exercise/shapes_game/index.html [hunk 1]
```diff
@@ -1,12 +1,12 @@
-<!DOCTYPE html>
+subl in <!DOCTYPE html>
```
Comment (tigarcia) [inline canvas_exercise/shapes_game/index.html]: Probably a mistake here

File: canvas_exercise/shapes_game/index.html [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
 <html>
 <head>
   <title>Shapes!!</title>
   <link rel="stylesheet" type="text/css" href="vendor/bootstrap.min.css">
   <link rel="stylesheet" type="text/css" href="styles.css">
   <script type="text/javascript" src="vendor/jquery-1.12.4.min.js"></script>
   <script type="text/javascript" src="vendor/bootstrap.min.js"></script>
-  <script type="text/javascript" src="index.js"></script>
+  <script type="text/javascript" src="index_doesntWork.js"></script>
```
Comment (tigarcia) [inline canvas_exercise/shapes_game/index.html]: Why did you add doesntwork?

File: recursion_exercise/recursion.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    if(arr.length === 1) return arr[0]
+    return arr[0] * productOfArray(arr.slice(1))
+}
+
+
+function collectStrings(obj) {
+    var newArr = [];
+
+    function collectStringsHelper(obj) {
+        for(var i=0; i<Object.keys(obj).length; i++) {
```
Comment (tigarcia) [inline recursion_exercise/recursion.js]: I would only call Object.keys(obj) once.  Save it to a variable and put it outside of the for loop.  Your code would then be a little more readable:

```
var keys = Object.keys(obj);
for (var i = 0; i < keys.length; i++) {
    if (typeof obj[keys[i]] === 'string') {
         newArr.push(obj[keys[i]]);
    else {
        return collectStringsHelper(obj[keys[i]]);
    }
}
```

File: recursion_exercise/recursion.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    collectStringsHelper(obj)
+
+    return newArr;
+}
+
+
+function contains(obj, value) {
+    for(var key in obj) {
+        if(typeof obj[key] != 'object') {
```
Comment (tigarcia) [inline recursion_exercise/recursion.js]: !==

File: recursion_exercise/recursion.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        }    
+    }
+    return false;
+}
+
+function search(arr, val) {
+    var i = 0
+
+    function searchHelper(arr, val) {
+        if(arr.length < 1) {
```
Comment (tigarcia) [inline recursion_exercise/recursion.js]: I prefer spaces after if:

```
if (arr.length < 1) {
```

File: recursion_exercise/recursion.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            i = -1;
+            return;
+        } 
+        if(arr[0]===val) {
```
Comment (tigarcia) [inline recursion_exercise/recursion.js]: Same here, space after if and also space between the ===:

```
if (arr[0] === val) {
```

File: recursion_exercise/recursion.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            return;
+        }
+        i += 1;
+        searchHelper(arr.slice(1), val);
+    }
+
+   searchHelper(arr, val);
+   return i;
+}
```
Comment (tigarcia) [inline recursion_exercise/recursion.js]: Don't forget a new line at the end of the file.  I've seen bugs caused by this, so I'm extra sensitive to it.

File: testing_exercise/testing.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
+function replaceWith(string, ltrOld, ltrNew) {
+	var newString = "";
+	for(var i=0; i<string.length; i++) {
+		if(string[i]==ltrOld) {
+			newString += ltrNew;
+		} else {
+			newString += string[i];
+		}
+	};
```
Comment (tigarcia) [inline testing_exercise/testing.js]: No semicolons at the end of for loops

File: testing_exercise/testing.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
+function replaceWith(string, ltrOld, ltrNew) {
+	var newString = "";
+	for(var i=0; i<string.length; i++) {
```
Comment (tigarcia) [inline testing_exercise/testing.js]: space after the for, space between = sign, space between the < sign:

for (var i = 0; i < string.length; i++) {

File: testing_exercise/testing.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+		if(string[i]==ltrOld) {
```
Comment (tigarcia) [inline testing_exercise/testing.js]: use === and a space:

```
if (string[i] === ltrOld) {
```

File: testing_exercise/testing.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+			newString += ltrNew;
+		} else {
+			newString += string[i];
+		}
+	};
+	if(newString === "") {
+		return string;
+	};
```
Comment (tigarcia) [inline testing_exercise/testing.js]: No semicolons at the end of if statements.

File: testing_exercise/testing.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+		newArr.push(arr[i % arr.length]);
+	}
+	return newArr;
+}
+
+//
+
+function acceptNumbersOnly(args) {
+	for(var i=0; i<arguments.length; i++) {
+		if(isNaN(arguments[i]))
```
Comment (tigarcia) [inline testing_exercise/testing.js]: I prefer to use { } around if statements even if it's one line.  If you want to make it a little cleaner, I'm ok with this:

```
if (isNaN(arguments[i]) { return false; }
```

File: testing_exercise/testing.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+	for(var i=0; i < arr.length*num; i++) {
+		newArr.push(arr[i % arr.length]);
+	}
+	return newArr;
+}
+
+//
+
+function acceptNumbersOnly(args) {
+	for(var i=0; i<arguments.length; i++) {
```
Comment (tigarcia) [inline testing_exercise/testing.js]: spaces!

File: testing_exercise/testing.js [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+		if(isNaN(arguments[i]))
+			return false;
+	}
+	return true;
+}
+
+//
+
+function mergeArrays(arr1, arr2) {
+	if(arr2 == undefined || arr2 == [])
```
Comment (tigarcia) [inline testing_exercise/testing.js]: triple equals, curly braces, spaces

File: testing_exercise/testing.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+		return arr1;
+	if(typeof arr1 != 'object' || typeof arr2 != 'object')
```
Comment (tigarcia) [inline testing_exercise/testing.js]: !==, curly braces, spaces

File: testing_exercise/testing.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+		return undefined;
+	for(var i=0; i<arr2.length; i++) {
+		arr1.push(arr2[i])
+	}
+	return arr1.sort(function(a, b) {
+		return a - b;
+		});
```
Comment (tigarcia) [inline testing_exercise/testing.js]: This closing curly and end of the sort should be unindented once

File: recursion_exercise/recursion.js (hunk 7 above)
Comment (amandanagai) [inline recursion_exercise/recursion.js]: What does this mean? Can you explain further? Thanks!

Comment (elie) [inline recursion_exercise/recursion.js]: At the end of each file - add a new line with nothing it it.

File: AJAXhackerNews.css [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+body {
+    background-color: #D5D5D5;
+}
+
+form {
+
+}
+
+.favsList {
+	display: none;
```
Comment (elie) [inline AJAXhackerNews.css]: watch your indentation - and remove the empty selector above

File: AJAXhackerNews.html [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+                  </div> -->
+            </div>
+            <div class="content">
+                <form id="login">
+                    <label>login</label>
+                    <br>
+                    <input class="usernameLogin" type="text" placeholder="username">  
+                    <br>
+
+                    <input class="passLogin" type="text" placeholder="password">
```
Comment (elie) [inline AJAXhackerNews.html]: use type="password" for passwords on your input

File: AJAXhackerNews.html [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+                    <input class="submitLogin" type="submit" value="Submit">
+                </form>
+                <br>
+                <form id="signup">
+                    <label>signup</label>
+                    <br>
+                    <input class="usernameSignup" type="text" placeholder="username">  
+                    <br>
+
+                    <input class="passSignup" type="text" placeholder="password">
```
Comment (elie) [inline AJAXhackerNews.html]: same

File: AJAXhackerNews.js [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+                    hacker_news_story: {
+                        by: storiesListObj[$(e.target).next().attr('class')].by, 
+                        story_id: storiesListObj[$(e.target).next().attr('class')].id,
+                        title: storiesListObj[$(e.target).next().attr('class')].title,
+                        url: storiesListObj[$(e.target).next().attr('class')].url
+                    }
+                })
+            }).then(function(res) {
+                $(e.target).data('deleteID', res.id)
+                debugger
```
Comment (elie) [inline AJAXhackerNews.js]: remove this debugger

File: call_apply_bind_exercise/callApplyBind.js [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+	this.fullName = fullName;
+	this.sayHi = function() {
+		return `This person's name is ${this.fullName}`
+	}
+}
+
+var obj = new  IntroducePerson('Harry Potter')
+
+
+// Answer: arguments and sets
```
Comment (elie) [inline call_apply_bind_exercise/callApplyBind.js]: not exactly - sets are not array-like-objects, they are just a type of object. HTML collections (what you get from $() and document.getElementsByTagName or querySelectorAll are array like objects)

File: call_apply_bind_exercise/callApplyBind.js [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+			} else if (guess > answer) {
+				return "You're too high!"
+			} else {
+				return "You're too low!"
+			}
+		} else if (guessCount === amount) {
+			guessCount += 1;
+			return `No more guesses the answer was ${answer}`
+		} else {
+			return 'You are all done playing!'
```
Comment (elie) [inline call_apply_bind_exercise/callApplyBind.js]: do we need this else?
